### PR TITLE
chore: Upgrade to Angular v14

### PR DESCRIPTION
**Pull request by rgant**
__DESCRIPTION:__
Angular v13 support window ends [2023-05-04](https://angular.io/guide/releases#actively-supported-versions). Angular v16 is expected [2023-05-01](https://angular.io/guide/releases#release-schedule) (BTW [documentation](https://developers.transifex.com/docs/angular-sdk) does not express support for Angular v15 currently. But the [code](https://github.com/transifex/transifex-javascript/blob/master/packages/angular/projects/tx-native-angular-sdk/package.json#L19-L21) does. Should this be updated?)
My analysis of the current Angular code shows nothing that is a breaking change from Angular v13 to v15.

Additionally I have refreshed the dependencies and updated the lint configuration to current recommendations from Angular / @angular-eslint/eslint-plugin. All code changes are automatic formatting and fixing from tooling.

Finally I have run `npx package-json-validator --warnings --recommendations` and `npx sort-package-json`.

__STEPS TO TEST:__
1. `ng lint`
2. `ng test`
3. `ng build`

All linting and tests pass on my system. Build appears to be good as well. All of my future contributions to this library will likely be based off of these changes.

### CODE REVIEW

**Comment by rgant** [comment]
Merging this branch will close #187

**Comment by pablotransifex** [comment]
@rgant I appreciate the effort you put in lintng and cleaning the package. We in the beginning decided to go with the Airbnb linting standards in general for the TXNative repository, but I actually was aware of the specific linting rules and standards in Angular development. I'll check these changes and let you know how we decided to proceed.

Again, thank you very much.

**Comment by pablotransifex** [comment]
@rgant It seems that the tests have some problem, in order to reproduce locally, please run in the root folde of the repo:

- npm run bootstrap

and in the angular folder:

- npm test

**Comment by rgant** [comment]
I'll look into the error, but best I can tell that is a build error and has less to do with my code and more to do with old versions of lerna and old versions of Angular (14 is still old) with brand new versions of NodeJS v18.

I tend to expect that Transifex isn't going to wholly accept my changes. Which is fine, but I did promise to suggest improvements before based on my company's experience using your tools. We have completely abandoned this Angular implementation because of problems with how it works. I will be completely re-writing everything there to work in (what I believe is) a more _Angular-y_ way. 

It's up to you if you want to wait to see the final results before reviewing, or review individually. I won't open any other pull requests here, but I will be recording them in my fork for you to look at if you desire.

**Comment by rgant** [comment]
I believe the issue is that the [library's package.json](https://github.com/rgant/transifex-javascript/blob/b2a8dc4a01de32e9ff1f785d962458e604c8cdfd/packages/angular/projects/tx-native-angular-sdk/package.json#L21) includes Angular 12 and 13. But I haven't had a chance yet to try and remove those.

**Comment by pablotransifex** [comment]
@rgant The problem was this part in the `package,json` inside the library:

```
  "dependencies": {
    "tslib": "^2.3.0"
  }
```

this generates a `node_modules` inside it that is not the same as in the root folder, and thus the discrepancy.

**Comment by rgant** [comment]
Good catch, I will add documentation about removing that since that is the default when Angular v14 creates a library.

File: packages/angular/projects/tx-native-angular-sdk/tests/T.decorator.spec.ts [hunk 1]
```diff
@@ -1,77 +1,55 @@
...
   it('should test the decorator T', () => {
-    const compiled = fixture.debugElement.nativeElement;
-    expect(tx.translate).toHaveBeenCalled();
-    expect((compiled as HTMLDivElement).innerHTML)
-      .toContain('ok-translated-dec');
+    const fixture = TestBed.createComponent(TestComponent);
+    const component = fixture.componentInstance;
+
+    fixture.detectChanges();
+    expect(tx.translate).toHaveBeenCalledTimes(2); // Why twice?
```
Comment (pablotransifex) [inline packages/angular/projects/tx-native-angular-sdk/tests/T.decorator.spec.ts:35]: I think we can ignore this, and use a simple `hasBeenCalled()` . I couldn't find any reason why it's called twice... probably some internal mechanism for decorators? Let's assume that is working fine.

File: packages/angular/projects/tx-native-angular-sdk/tests/T.decorator.spec.ts [hunk 2]
```diff
@@ -1,77 +1,55 @@
...
-    expect(tx.translate).toHaveBeenCalled();
-    setTimeout(function () {
-      expect((compiled as HTMLDivElement).innerHTML)
-        .toContain('ok-translated-dec');
-    }, 1000);
+    const fixtureWithInstance = TestBed.createComponent(TestWithInstanceComponent);
+    const component = fixtureWithInstance.componentInstance;
+
+    fixtureWithInstance.detectChanges();
+    expect(tx.translate).toHaveBeenCalledTimes(2); // Why twice?
```
Comment (pablotransifex) [inline packages/angular/projects/tx-native-angular-sdk/tests/T.decorator.spec.ts:48]: I think we can ignore this, and use a simple `hasBeenCalled()` . I couldn't find any reason why it's called twice... probably some internal mechanism for decorators? Let's assume that is working fine.

File: packages/angular/projects/tx-native-angular-sdk/src/lib/language-picker.component.ts [hunk 3]
```diff
@@ -50,64 +48,49 @@ export class LanguagePickerComponent implements OnInit, OnDestroy {
...
+    });
   }
 
   ngOnInit(): void {
-    // Do not retrieve languages in the initialization
-    // if alternative instace found, will fetch languages
+    // Do not retrieve languages in the initialization if alternative instance found, will fetch languages
     // when the instance is ready using a subscription
-    if (this.instance && this.instance.alias) {
+    if (this.instance.alias) {
```
Comment (pablotransifex) [inline packages/angular/projects/tx-native-angular-sdk/src/lib/language-picker.component.ts:70]: Maybe use `this.instance?.alias` just to be sure?

**Review by pablotransifex — COMMENTED** [review]
@rgant Looks good, but I want to test with older versions of Angular. Did you test in a project with v12/13 of Angular?

**Comment by pablotransifex** [comment]
@rgant I have errors when compiling in Angular 12 and 13, like this one:
image

using this PR is breaking those versions, is this expected?

**Comment by rgant** [comment]
Angular 12 and 13 are no longer supported versions of Angular: https://angular.io/guide/releases#actively-supported-versions

**Comment by pablotransifex** [comment]
@rgant correct, I was thinking in creating a new release with the v14+ for angular and keep another release for upto v13. This way we can support older versions with the current implementation, and improve the new installations with your refactoring (when ready).

From my side there is no problem, I would like to discuss this and back to you.

**Comment by pablotransifex** [comment]
@rgant correct, I was thinking in creating a new release with the v14+ for angular and keep another release for upto v13. This way we can support older versions with the current implementation, and improve the new installations with your refactoring (when ready).

What do you think?

**Comment by rgant** [comment]
That is one way. I think a more typical way is to say that users of old versions of frameworks should install old versions of add-ons.

**Comment by egilkh** [comment]
I agree with rgrant. And hope this can be merged :)

**Comment by pablotransifex** [comment]
> I agree with rgrant. And hope this can be merged :)

@egilkh Due to vacations we cannot do it now, but we will as soon as possible.

**Comment by pablotransifex** [comment]
@rgant Can you please rebase and resolve conflict in this PR in order to move on and release?

File: packages/angular/package.json [hunk 4]
```diff
@@ -40,35 +52,41 @@
...
+    "karma-coverage": "~2.2.0",
+    "karma-jasmine": "~5.1.0",
+    "karma-jasmine-html-reporter": "~2.0.0",
+    "ng-packagr": "^14.2.2",
     "ts-node": "~8.3.0",
-    "tslint": "~6.1.0",
-    "typescript": "~4.4.4"
+    "typescript": "~4.7.4"
+  },
+  "engines": {
```
Comment (pablotransifex) [inline packages/angular/package.json:88]: @rgant We need to remove this, I think.

**Review by pablotransifex — COMMENTED** [review]
@rgant I don't know what changed since last time, but there are errors in the compilation, related to angular dependencies, also notices that you added node and npm versions to package.json, that should not be there.

Last time we build this PR was good, I don't know what changed and now is breaking.

**Comment by pablotransifex** [comment]
@rgant I don't know what changed since last time, but there are errors in the compilation, related to angular dependencies, also notices that you added node and npm versions to package.json, that should not be there.

Last time we build this PR was good, I don't know what changed and now is breaking.

**Comment by pablotransifex** [comment]
@rgant Closing this, we have fixed the build on another PR already merged. Feel free to let us know any other improvements that we can apply for the Angular 14 version we will release soon. And thank you very much again for the effort.

**Pull request closed without merging**